src/foolscap/connections/tor.py

```diff
@@ -0,0 +1,153 @@
+
+from __future__ import print_function, unicode_literals
+import time
+from twisted.internet.interfaces import IStreamClientEndpoint
+from twisted.internet.defer import inlineCallbacks, returnValue
+from twisted.internet.error import ConnectError
+import txtorcon
+import ipaddress
+from .timing import DebugTiming
+from .transit import allocate_tcp_port
+from .. import observer
+
+
+from zope.interface import implementer
+from twisted.internet import endpoints
+from foolscap.ipb import IConnectionHintHandler, InvalidHintError
+import txtorcon
+from .tcp import DOTTED_QUAD_RESTR, DNS_NAME_RESTR
+from .socks import SOCKS5
+
+def is_non_public_numeric_address(host):
+    # for numeric hostnames, skip RFC1918 addresses, since no Tor exit
+    # node will be able to reach those. Likewise ignore IPv6 addresses.
+    try:
+        a = ipaddress.ip_address(host)
+    except ValueError:
+        return False # non-numeric, let Tor try it
+    if a.version != 4:
+        return True # IPv6 gets ignored
+    if (a.is_loopback or a.is_multicast or a.is_private or a.is_reserved
+        or a.is_unspecified):
+        return True # too weird, don't connect
+    return False
+
+HINT_RE = re.compile(r"^(tcp|tor):(%s|%s):(\d+){1,5}$" % (DOTTED_QUAD_RESTR,
+                                                          DNS_NAME_RESTR))
+
+@implementer(IConnectionHintHandler)
+class _Common:
+    # subclasses must:
+    #  define _connect()
+    #  set self._socks_hostname, self._socks_port
+
+    def __init__(self):
+        self._connected = False
+        self._when_connected = observer.OneShotObserverList()
+
+    def _maybe_connect(self, reactor):
+        if not self._connected:
+            self._connected = True
+            # connect
+            d = self._connect()
+            d.addBoth(self._when_connected.fire)
+        return self._when_connected
+
+    @inlineCallbacks
+    def hint_to_endpoint(self, hint, reactor):
+        # Return (endpoint, hostname), where "hostname" is what we pass to the
+        # HTTP "Host:" header so a dumb HTTP server can be used to redirect us.
+        mo = HINT_RE.search(hint)
+        if not mo:
+            raise InvalidHintError("unrecognized TCP/Tor hint")
+        host, port = mo.group(1), int(mo.group(2))
+        if is_non_public_numeric_address(host):
+            raise InvalidHintError("ignoring non-Tor-able ipaddr %s" % host)
+        yield self._maybe_connect()
+        # txsocksx doesn't like unicode: it concatenates some binary protocol
+        # bytes with the hostname when talking to the SOCKS server, so the
```

Comment:
Actually also txsocksx doesn't (and probably won't) support Python3, either :/

This led me to write a custom socks implementation in txtorcon, so this will go away once txtorcon 1.0 is available. Also, it supports the two Tor extra SOCKS opcodes.